spec/type_references_and_ownership.md

```diff
@@ -0,0 +1,331 @@
+         owning resource relationships in Azure are this kind.
+
+
+Note that sometimes an owning type has its dependent type embedded as a property 
+(for example: `RouteTable` has the property [RouteTablePropertiesFormat](https://schema.management.azure.com/schemas/2020-03-01/Microsoft.Network.json)).
+Most types do not embed the dependent resource directly in the owning resource. We need to cater to both scenarios.
+
+## What does expressing these relationships look like in existing solutions?
+
+This section examines how other operator solutions have tackled these problems. We look at:
+- ARM templates
+- AzureServiceOperator (ASO)
+- k8s-infra
+
+### Related/Linked resources
+
+#### What ARM does
+These are just properties (often but not always called `id`) which refer to the fully qualified ARM ID of another resource. For example see a
+[sample deployment template for a VMSS refering to an existing vnet](https://github.com/Azure/azure-quickstart-templates/blob/master/201-vmss-existing-vnet/azuredeploy.json#L136).
+
+```json
+"properties": {
+  "subnet": {
+    "id": "[resourceId(parameters('existingVnetResourceGroupName'), 'Microsoft.Network/virtualNetworks/subnets', parameters('existingVnetName'), parameters('existingSubNetName'))]"
+  },
+}
+```
+
+#### What ASO does
+Similar to what ARM templates do, ASO uses the decomposition of fully qualified resource id to reference another resource, as seen [here for VMSS -> VNet](https://github.com/Azure/azure-service-operator/blob/92240406aff3863f3a267d8a1dc1e28aa3e841ae/api/v1alpha1/azurevmscaleset_types.go#L25)
+
+```go
+type AzureVMScaleSetSpec struct {
+    ...
+	Location               string `json:"location"`
+	ResourceGroup          string `json:"resourceGroup"`
+	VirtualNetworkName     string `json:"virtualNetworkName"`
+	SubnetName             string `json:"subnetName"`
+    ...
+}
+```
+
+These properties are combined into a fully qualified ARM id like so:
+
+```go
+subnetIDInput := helpers.MakeResourceID(
+    client.SubscriptionID,
+    resourceGroupName,
+    "Microsoft.Network",
+    "virtualNetworks",
+    vnetName,
+    "subnets",
+    subnetName,
+)
+```
+
+Currently ASO does not support cross-subscription references (and some of the resources such as VMSS above don't even support cross-resource group references), but it in theory could by adding some more parameters.
+
+#### What k8s-infra does
+k8s-infra is a bit different in that resource references are in Kubernetes language (namespace + name) and not Azure language (resource group + resource name).
+All resource references are done using a special type `KnownTypeReference` which contains the fully qualified Kubernetes name for the resource.
+
+### Dependent Resources
+#### What ARM does
+ARM template deployments support [two different ways of deploying dependent resources](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/child-resource-name-type):
```

Comment:
I've experimented with these 3 ways of deploying dependent resources. There might even be more ways...

Sibling level resources with a dependsOn:
```json
"resources": [
  {
    "apiVersion": "2019-09-01",
    "type": "Microsoft.Network/virtualNetworks",
    "name": "foo",
    "location": "[resourceGroup().location]",
    "properties": {
      "addressSpace": {
        "addressPrefixes": [
          "10.0.0.0/16"
        ]
      }
    }
  },
  {
    "apiVersion": "2019-09-01",
    "type": "Microsoft.Network/virtualNetworks/subnets",
    "name": "foo/foo-subnet",
    "properties": {
      "addressPrefix": "10.0.0.0/24"
    },
    "dependsOn": [
      "Microsoft.Network/virtualNetworks/foo"
    ]
  }
]
```

Embedded within a set of child resources:
```json
"resources": [
  {
    "apiVersion": "2019-09-01",
    "type": "Microsoft.Network/virtualNetworks",
    "name": "foo",
    "location": "[resourceGroup().location]",
    "properties": {
      "addressSpace": {
        "addressPrefixes": [
          "10.0.0.0/16"
        ]
      }
    },
    "resources": [
      {
        "apiVersion": "2019-09-01",
        "type": "subnets",
        "name": "foo-subnet",
        "properties": {
          "addressPrefix": "10.0.0.0/24"
        },
        "dependsOn": [
          "Microsoft.Network/virtualNetworks/foo"
        ]
      }
    ]
  }
]
```

Embedded directly as an object within the owning resource:
```json
"resources": [
  {
    "apiVersion": "2019-09-01",
    "type": "Microsoft.Network/virtualNetworks",
    "name": "foo",
    "location": "[resourceGroup().location]",
    "properties": {
      "addressSpace": {
        "addressPrefixes": [
          "10.0.0.0/16"
        ]
      },
      "subnets": [
        {
          "name": "foo-subnet",
          "properties": {
            "addressPrefix": "10.0.0.0/24"
          }
        }
      ]
    }
  }
]
```

Reply:
Ah, yeah -- the 3rd option doesn't always work (it depends on the resource type involved). Some list their dependent resource types as a property (in this case `subnets`), but often they are not listed as a property on the parent type.

I didn't include the third option in the list because it's not really applicable to all resources, whereas the other two ways are.